### PR TITLE
typeguard fixes for beautiful-mnsit JIT=2 METAL=1 [run_process_replay]

### PR DESCRIPTION
**Pull request by greg-niemeyer**
[geohot:](https://discord.com/channels/1068976834382925865/1068982781490757652/1270028429491699734)
> woah
import typeguard
typeguard.install_import_hook()
is the runtime type checker i've been looking for
hit this typeguard.TypeCheckError: argument "global_size" (list) is not a tuple installing it and running beautiful mnist
if someone wants to fix these things up and submit PRs would be great. ideally find out why the static type checker misses them

This PR gets `JIT=<0,1,2> METAL=1 python3 examples/beautiful_mnist.py` passing with `typeguard.install_import_hook()`.

As to the question of 'why the static type checker misses them' -- the overlooked type inconsistency in do_reduce() is because the function wasn't annotated, so I added the missing annotation. I have theories about why the static type checker missed the others, but they're unverified and mostly boil down to some variation of 'too complicated/dynamic for the static type checker'.

### CODE REVIEW

File: tinygrad/codegen/uopgraph.py [hunk 1]
```diff
@@ -396,11 +396,12 @@ def do_expand(root:UOp):
   return UOp(UOps.EXPAND, root.dtype, tuple(new_srcs), expand_args)
 
 acc_number = 0
-def do_reduce(root):
+def do_reduce(root:UOp):
   global acc_number
-  reduce_parented, reduce_unparented = partition(root.src[1:], lambda x: x in root.src[0].parents)
+  reduce_parented, reduce_unparented = partition(list(root.src[1:]), lambda x: x in root.src[0].parents)
```
Comment (chenyuxyz) [inline tinygrad/codegen/uopgraph.py:401]: change partition function to Iterable instead of change the user code

Comment (greg-niemeyer) [reply tinygrad/codegen/uopgraph.py]: done

**Comment by chenyuxyz** [comment]
is this good to merge (it's draft)?

**Comment by greg-niemeyer** [comment]
> is this good to merge (it's draft)?

sorry, yeah -- updated

**Pull request merged**